### PR TITLE
AL-377 Disco deploy honor ami option for update test 012317

### PR DESCRIPTION
**Pull request by rruster**

### CODE REVIEW

File: test/unit/test_disco_deploy.py [hunk 1]
```diff
@@ -1133,7 +1242,7 @@ def test_run_integration_tests_success(self):
         self._ci_deploy._get_old_instances = MagicMock(return_value=[inst1])
         self._ci_deploy._get_new_instances = MagicMock(return_value=[inst2])
         self.assertIsNone(self._ci_deploy.test_ami(ami, dry_run=False))
-        self._ci_deploy._get_old_instances.assert_called_with(ami.id)
+        self._ci_deploy._get_old_instances.assert_called_with(ami.id, launch_time=ANY)
```
Comment (jconstance-amplify) [inline test/unit/test_disco_deploy.py]: I much prefer the specificity of saying `launch_time=ANY` as opposed to just `ANY`. Can we do that for all of these calls?

File: test/unit/test_disco_deploy.py [hunk 2]
```diff
@@ -1273,6 +1382,38 @@ def test_test_wo_amis(self):
...
+        self._ci_deploy._disco_bake.list_amis = MagicMock(return_value=[])
+        self._ci_deploy.get_test_amis = MagicMock(return_value=[])
+        self._ci_deploy.test_ami = MagicMock()
+        self._ci_deploy.test()
+        self.assertEqual(self._ci_deploy.get_test_amis.call_count, 0)
+        self._ci_deploy._disco_bake.list_amis.assert_called_with(ami_ids=[self._amis_by_name['mhcbar 2'].id])
+        self.assertEqual(self._ci_deploy.test_ami.call_count, 0)
+
+    def test_test_wo_restrict_ami(self):
+        '''Test test with spcified restrict amis'''
```
Comment (jconstance-amplify) [inline test/unit/test_disco_deploy.py]: This docstring doesn't seem to agree with the name of the method. Also, wouldn't we want to verify that it uses the latest ami?

File: test/unit/test_disco_deploy.py [hunk 3]
```diff
@@ -1273,6 +1382,38 @@ def test_test_wo_amis(self):
...
+        self._ci_deploy._disco_bake.list_amis = MagicMock(return_value=amis)
+        self._ci_deploy.get_test_amis = MagicMock(return_value=[])
+        self._ci_deploy.test_ami = MagicMock()
+        self._ci_deploy.test()
+        self.assertEqual(self._ci_deploy.get_test_amis.call_count, 0)
+        self._ci_deploy._disco_bake.list_amis.assert_called_with(ami_ids=[self._amis_by_name['mhcbar 2'].id])
+        self.assertEqual(self._ci_deploy.test_ami.call_count, 1)
+
+    def test_test_with_invalid_restrict_ami(self):
+        '''Test test with spcified restrict amis'''
```
Comment (jconstance-amplify) [inline test/unit/test_disco_deploy.py]: `spcified` -> `specified`

File: test/unit/test_disco_deploy.py [hunk 4]
```diff
@@ -1273,6 +1382,38 @@ def test_test_wo_amis(self):
         self._ci_deploy.test()
         self.assertEqual(self._ci_deploy.test_ami.call_count, 0)
 
+    def test_test_with_restrict_ami(self):
+        '''Test test with spcified restrict amis'''
```
Comment (jconstance-amplify) [inline test/unit/test_disco_deploy.py]: `spcified` -> `specified`

File: test/unit/test_disco_deploy.py [hunk 5]
```diff
@@ -1286,6 +1427,38 @@ def test_update_wo_amis(self):
...
+        self._ci_deploy._disco_bake.list_amis = MagicMock(return_value=[])
+        self._ci_deploy.get_update_amis = MagicMock(return_value=[])
+        self._ci_deploy.update_ami = MagicMock()
+        self._ci_deploy.update()
+        self.assertEqual(self._ci_deploy.get_update_amis.call_count, 0)
+        self._ci_deploy._disco_bake.list_amis.assert_called_with(ami_ids=[self._amis_by_name['mhcbar 2'].id])
+        self.assertEqual(self._ci_deploy.update_ami.call_count, 0)
+
+    def test_update_wo_restrict_ami(self):
+        '''Test test with spcified restrict amis'''
```
Comment (jconstance-amplify) [inline test/unit/test_disco_deploy.py]: Docstring and method mismatch.

File: disco_aws_automation/disco_aws.py [hunk 6]
```diff
@@ -573,9 +574,19 @@ def instances_from_hostclasses(self, hostclasses):
             if instance.tags.get("hostclass", "-") in hostclasses
         ]
 
-    def instances_from_amis(self, ami_ids):
-        """Returns instances matching any of a list of AMI ids"""
-        return self.instances(filters={"image_id": ami_ids})
+    def instances_from_amis(self, ami_ids, group_name=None, create_date=None):
```
Comment (rruster) [inline disco_aws_automation/disco_aws.py]: Done

File: disco_aws_automation/disco_aws.py [hunk 7]
```diff
@@ -573,9 +574,19 @@ def instances_from_hostclasses(self, hostclasses):
...
+        """Returns instances matching any of a list of AMI ids and filtered by ASG name and created date"""
```
Comment (rruster) [inline disco_aws_automation/disco_aws.py]: done

File: disco_aws_automation/disco_aws_util.py [hunk 8]
```diff
@@ -27,6 +28,15 @@ def get_tag_value(tag_list, key):
     return None
 
 
+def get_instance_launch_time(instance):
+    """
+    extract the launch time fro the instance and return a datetime object
```
Comment (rruster) [inline disco_aws_automation/disco_aws_util.py]: Done

File: disco_aws_automation/disco_aws_util.py [hunk 9]
```diff
@@ -27,6 +28,15 @@ def get_tag_value(tag_list, key):
...
+    :param instance:
```
Comment (rruster) [inline disco_aws_automation/disco_aws_util.py]: Done

File: disco_aws_automation/disco_deploy.py [hunk 10]
```diff
@@ -122,6 +124,11 @@ def get_latest_failed_amis(self):
         '''Returns latest failed AMI for each hostclass'''
         return self.get_latest_ami_in_stage_dict('failed')
 
+    def get_amis_by_ids(self):
+        '''Returns the AMI Objects for the specified ami_ids stored in the config variable restrict_amis'''
+        if self._restrict_amis is not None:
```
Comment (rruster) [inline disco_aws_automation/disco_deploy.py]: I'm going to remove this function as it's not really useful.

File: disco_aws_automation/disco_deploy.py [hunk 11]
```diff
@@ -288,18 +298,26 @@ def handle_nodeploy_ami(self, ami, pipeline_dict=None, dry_run=False, old_group=
         if ami_test_failed:
             raise RuntimeError("Smoke test for non-deploy Hostclass %s AMI %s failed", hostclass, ami.id)
 
-    def _get_old_instances(self, new_ami_id):
-        '''Returns instances of the hostclass of new_ami_id that are not running new_ami_id'''
+    def _get_old_instances(self, new_ami_id, launch_time=None):
+        '''Returns instances of the hostclass of new_ami_id that are not running new_ami_id
+        or using launch date
```
Comment (rruster) [inline disco_aws_automation/disco_deploy.py]: Done

File: disco_aws_automation/disco_deploy.py [hunk 12]
```diff
@@ -288,18 +298,26 @@ def handle_nodeploy_ami(self, ami, pipeline_dict=None, dry_run=False, old_group=
...
         all_instances = self._disco_aws.instances(instance_ids=all_ids)
-        return [inst for inst in all_instances if inst.image_id != new_ami_id]
+        return [inst for inst in all_instances
+                if (inst.image_id != new_ami_id) or
+                (launch_time and get_instance_launch_time(inst) < launch_time)]
 
-    def _get_new_instances(self, new_ami_id):
-        '''Returns instances running new_ami_id'''
+    def _get_new_instances(self, new_ami_id, launch_time=None):
+        '''Returns instances running new_ami_id
```
Comment (rruster) [inline disco_aws_automation/disco_deploy.py]: Done

File: disco_aws_automation/disco_deploy.py [hunk 13]
```diff
@@ -288,18 +298,26 @@ def handle_nodeploy_ami(self, ami, pipeline_dict=None, dry_run=False, old_group=
...
+        If launch_date is specify, select only instance created after the specified date
```
Comment (rruster) [inline disco_aws_automation/disco_deploy.py]: Done

File: disco_aws_automation/disco_deploy.py [hunk 14]
```diff
@@ -402,10 +428,10 @@ def handle_tested_ami(self, ami, pipeline_dict=None, run_tests=False, dry_run=Fa
         old_ami_id = self._get_latest_other_image_id(ami.id)
         if old_ami_id:
             post_hostclass_dict["ami"] = old_ami_id
-        else:
+        elif old_ami_id != ami:
```
Comment (rruster) [inline disco_aws_automation/disco_deploy.py]: My test is incorrect and not necessary

File: disco_aws_automation/disco_deploy.py [hunk 15]
```diff
@@ -344,6 +362,11 @@ def handle_tested_ami(self, ami, pipeline_dict=None, run_tests=False, dry_run=Fa
         ami_test_failed = False
         hostclass = DiscoBake.ami_hostclass(ami)
 
+        try:
+            old_ami_id = self._disco_aws.instance_from_hostname(hostclass).image_id
```
Comment (rruster) [inline disco_aws_automation/disco_deploy.py]: Not necessary, removing

File: test/unit/test_disco_deploy.py [hunk 16]
```diff
@@ -326,6 +329,17 @@ def test_get_test_amis_from_pipeline(self):
...
+    def test_get_amis_by_ids_with_restrict_ami(self):
+        '''Tests that we can get amis by id using the restrict_amis option'''
+        self._ci_deploy._restrict_amis = [self._amis_by_name['mhcbar 2'].id]
+        amis = [self._amis_by_name['mhcbar 2']]
+        self._ci_deploy._disco_bake.get_amis = MagicMock(return_value=amis)
+        self.assertEquals(self._ci_deploy.get_amis_by_ids(), [self._amis_by_name['mhcbar 2']])
+
+    def test_get_amis_by_ids_no_restrict_ami(self):
+        '''Tests that we can get amis by id without any the restrict_amis option'''
+        self.assertIsNone(self._ci_deploy.get_amis_by_ids())
```
Comment (rruster) [inline test/unit/test_disco_deploy.py]: I removed this function, will remove the unit test too

File: test/unit/test_disco_deploy.py (hunk 1 above)
Comment (rruster) [inline test/unit/test_disco_deploy.py]: 👍

File: test/unit/test_disco_deploy.py (hunk 5 above)
Comment (rruster) [inline test/unit/test_disco_deploy.py]: Done

File: disco_aws_automation/disco_deploy.py [hunk 17]
```diff
@@ -814,20 +841,34 @@ def update_ami(self, ami, dry_run, deployment_strategy=None):
             )
 
     def test(self, dry_run=False, deployment_strategy=None):
-        '''Tests a single untested AMI and marks it as tested or failed'''
-        amis = self.get_test_amis()
+        '''
+        Tests a single AMI and marks it as tested or failed.
+        If the ami id is specify using the option --ami then run test on the specified ami
```
Comment (aoskotsky-amplify) [inline disco_aws_automation/disco_deploy.py]: Maybe should be `specified` instead of `specify` here

File: disco_aws_automation/disco_aws.py [hunk 18]
```diff
@@ -717,7 +736,8 @@ def wait_for_autoscaling_instances(self, metadata_list, timeout=AUTOSCALE_TIMEOU
 
         return self.instances(instance_ids=instance_ids) if instance_ids else []
 
-    def wait_for_autoscaling(self, ami_id, min_count, timeout=AUTOSCALE_TIMEOUT):
+    def wait_for_autoscaling(self, ami_id, min_count, timeout=AUTOSCALE_TIMEOUT, group_name=None,
```
Comment (jconstance-amplify) [inline disco_aws_automation/disco_aws.py]: Same here. If I provide both filters, I'd expect both to be used.

File: test/unit/test_disco_aws.py [hunk 19]
```diff
@@ -800,3 +803,68 @@ def test_is_running_termianted(self, mock_config, **kwargs):
...
+    def test_instances_from_amis_with_group_name(self, mock_config, **kwargs):
+        '''test get instances using ami ids in a specified group name'''
+        aws = DiscoAWS(config=mock_config, environment_name=TEST_ENV_NAME)
+        instances = [{"InstanceId": "i-123123aa"}]
+        aws.instances_from_asgs = MagicMock(return_value=instances)
+        self.assertEqual(aws.instances_from_amis('ami-12345678', group_name='test_group'), instances)
+        aws.instances_from_asgs.assert_called_with(['test_group'])
+
+    @patch_disco_aws
+    def test_instances_from_amis_with_create_date(self, mock_config, **kwargs):
```
Comment (jconstance-amplify) [inline test/unit/test_disco_aws.py]: `create_date` or `launch_time`? :)

File: disco_aws_automation/disco_aws.py [hunk 20]
```diff
@@ -573,9 +574,27 @@ def instances_from_hostclasses(self, hostclasses):
             if instance.tags.get("hostclass", "-") in hostclasses
         ]
 
-    def instances_from_amis(self, ami_ids):
-        """Returns instances matching any of a list of AMI ids"""
-        return self.instances(filters={"image_id": ami_ids})
+    def instances_from_amis(self, ami_ids, group_name=None, launch_time=None):
```
Comment (rruster) [inline disco_aws_automation/disco_aws.py]: Done

File: disco_aws_automation/disco_aws.py (hunk 18 above)
Comment (rruster) [inline disco_aws_automation/disco_aws.py]: Done

File: disco_aws_automation/disco_aws.py [hunk 21]
```diff
@@ -587,11 +587,13 @@ def instances_from_amis(self, ami_ids, group_name=None, launch_time=None):
         :return: List of instances.
         """
         if group_name:
-            return self.instances_from_asgs([group_name])
-        elif launch_time is None:
-            return self.instances(filters={"image_id": ami_ids})
+            instances = self.instances_from_asgs([group_name])
         else:
             instances = self.instances(filters={"image_id": ami_ids})
```
Comment (jconstance-amplify) [inline disco_aws_automation/disco_aws.py]: What if I wanted instances by both group_name and ami_id?

Comment (rruster) [inline disco_aws_automation/disco_aws.py]: Fixed

**Pull request merged**